### DISCUSSION TITLE
Stored procs based on scope

### DISCUSSION THREAD

**Discussion by slagtejn** [Mimetis/Dotmim.Sync#789] [Q&A]
Currently the stored procs are unique based on table and scope name. In our scenario we create ‘instances’ of the base scope for filters. This causes the same stored procs to be created  many times with the only difference being the scope name.  This creates so many stored procs that are not needed.  It also provisions an unnecessary amount of times.

Is there a solution for this? Sync framework had a scope config table and the scope info table linked to a config which solved this issue.  You could have multiple instances of the same scope but only one set of stored procs.

**Comment by Mimetis** [comment]
In `DMS`, creating a different scope is needed if:
- You are adding or removing tables
- You are adding or removing columns in a table
- You are adding or removing filters

If you are creating a new scope, that means one or more stored procedures are different (if adding / removing columns or filters) and / or one or more stored procedures are added / removed (if adding / removing tables)

Otherwise, no need to create a new scope.

That's being said, how would you implement such feature in DMS, especially the code part?
DMS is basically code first, a huge difference with Sync Framework.
Regarding the scopes tables in the databases (server / client) how will you do the implementation?

This is a very interesting question, that would imply a lot of modification in DMS to be implemented.
I don't know if it's worth it, but let's continue the discussion and see if we can find a solution

**Reply by slagtejn** [reply]
Yes most implementations will only have a few scopes and will follow the criteria you specified.  In our scenario we filter data based on users and it's a dynamically changing parameter.  Most implementations for this will have 1 scope with a filter.  The problem with this approach is when the parameter changes there is not a way to get the historical data without re-initializing the entire scope.  The amount of data re-syncing would be way to much.  This is why we create a scope per user so we can track the timestamp per filter parameter.  If we have 10k users in the database it will eventually create around 50k stored procs per table and 10k scope_info records with the same setup/schema.  The only difference being the scope name.

I am not sure how to implement this, but maybe there is a scope_config table which contains the scopes needed for provision and the scope_info table has a scope_config_id so you can have multiple instances of the same scope.  Provisioning would be based on the scope_config table.  This is how the tables were setup in sync framework.

Another option is to pass in an instance name when syncing a scope.  The instance name would be used to look up the scope info meta data like timestamp & last sync times.

**Reply by Mimetis** [reply]
I don't get it
 
> This is why we create a scope per user so we can track the timestamp per filter parameter.

You have, for each client scope (ie client) the `[scope_last_sync_timestamp]` value in the `[scope_info_history]` table.

Can you explain a little bit more how you are handling this scenario? 
A client database with different parameters value?

- Client Database 1: Parameter value 1: Sync from 0 to N
- Client Database 1: Parameter value 2: Sync ? from when ? to when ?

The problem is not the number of scopes, but probably because you need to create one scope per parameter Value, that is not something that should happens.

**Reply by slagtejn** [reply]
Yes we are creating one scope per parameter value to work around the issue of tracking the timestamp per parameter value.  If we don't create one scope per parameter how do you solve the case of ONLY syncing the delta data if the parameters change?  I feel like other user of this library would have this issue or they just haven't noticed it yet.  Basically if any filter parameter value changes on a scope there is no way to just get the delta data or historic data because it's tracked at the scope level (unless you re-initialize).  Since we are creating a scope per parameter value we are able to track that information and it's able to sync only the data needed and the historical data.

In your example...

- Client Database 1: Parameter value 1: Sync from 0 to N.  We create a scope 'ParameterData_1'.  The filter parameter value is 1
- Client Database 1: Parameter value 2: Sync ? from when ? to when ?.  We create a scope 'ParameterData_2'.  The filter parameter value is 2

Now if 'Client Database 1: Parameter value 3' is added a few weeks later we would create a scope 'ParameterData_3' which contains parameter value 3 so now we only get data related to parameter value 3 including all historical data.

**Reply by Mimetis** [reply]
Did you imagine having one client database per parameter value?
- Client 1 database 1: Parameter value 1
- Client 1 database 2: Parameter value 2

**Reply by slagtejn** [reply]
Yes only 1 database.  There won't be multiple databases per parameter value.

**Comment by Mimetis** [comment]
I think about your problem.
Can it be a solution if we are able to "redirect" a call to a specific stored procedure, instead of the "normal" call ?
We have an interceptor called `OnGetCommand`, that is called each time a command is created.
We can intercept this call, and change the stored procedure name.

Here is an example:

In this example, I'm creating 2 scopes on the server:
- First one is called "**default**" and contains all tables without filters.
- Second one is called "**filterproducts**" and contains all tables & filters.

The trick here is that we will **NOT** create all stored procedures for scope **filterproducts**:

``` cs
// Creating one Setup
// This Setup contains all tables without filters
// When provision, it will be called "default"
var setup = new SyncSetup("ProductCategory", "Product", "Address", 
    "Customer", "CustomerAddress");

var serverProvider = new SqlSyncProvider(DBHelper.GetDatabaseConnectionString(serverDbName));
var remoteOrchestrator = new RemoteOrchestrator(serverProvider);

// Provision this "default" scope
var serverScopeInfo = await remoteOrchestrator.ProvisionAsync("default", setup);

// Now, we are creating a new setup with a filter on ProductCategory and Product
var setup2 = new SyncSetup("ProductCategory", "Product", "Address", 
    "Customer", "CustomerAddress");
setup2.Filters.Add("ProductCategory", "Name");

var productFilter = new SetupFilter("Product");
productFilter.AddParameter("Name", DbType.String);
productFilter.AddJoin(Join.Left, "ProductCategory")
    .On("Product", "ProductCategoryID", "ProductCategory", "ProductCategoryID");
productFilter.AddWhere("Name", "ProductCategory", "Name");
setup2.Filters.Add(productFilter);

// Instead of Provisioning everything, we are 
// Provisionning the Stored procedures specific to filters and then save this new Scope
// Create a new ServerScopeInfo instance with schema and setup containing the filter
var schema2 = await remoteOrchestrator.GetSchemaAsync(setup2);
serverScopeInfo.Schema = schema2;
serverScopeInfo.Setup = setup2;
serverScopeInfo.Name = "filterproducts";

// Only generate the get changes with filters 
await remoteOrchestrator.CreateStoredProcedureAsync(serverScopeInfo, 
    "ProductCategory", default, DbStoredProcedureType.SelectChangesWithFilters, true);
await remoteOrchestrator.CreateStoredProcedureAsync(serverScopeInfo, 
    "ProductCategory", default, DbStoredProcedureType.SelectInitializedChangesWithFilters, true);

await remoteOrchestrator.CreateStoredProcedureAsync(serverScopeInfo, 
    "Product", default, DbStoredProcedureType.SelectChangesWithFilters, true);
await remoteOrchestrator.CreateStoredProcedureAsync(serverScopeInfo, 
    "Product", default, DbStoredProcedureType.SelectInitializedChangesWithFilters, true);

await remoteOrchestrator.SaveServerScopeInfoAsync(serverScopeInfo);

```

Now if you look at what have been generated on the server, you have something like this:

<img width="273" alt="image" src="https://user-images.githubusercontent.com/4592555/185612929-1fe1ceae-6838-48dc-ba8b-ba147d3aa8d6.png">

As you can see, all stored procedures for scope "**default**" have been generated, but only a few have been created for scope "**filterproducts**"

Now, we need to redirect all call to the good stored procedure.

> From the client perspective, since you will never have too many scopes, I don't do anything particularly...

``` cs
var clientProvider = new SqlSyncProvider(DBHelper.GetDatabaseConnectionString(clientDbName));

var agent = new SyncAgent(clientProvider, serverProvider);

// Eeach time a command is called on remote side, 
// We are making a redirection to "default" stored procedures, except for filters specific stored proc
agent.RemoteOrchestrator.OnGetCommand(args =>
{
    if (args.Command.CommandType == CommandType.StoredProcedure)
    {
        switch (args.CommandType)
        {
            case DbCommandType.SelectInitializedChangesWithFilters:
            case DbCommandType.SelectChangesWithFilters:
                break;
            default:
                args.Command.CommandText = args.Command.CommandText.Replace("_filterproducts_", "_default_");
                break;
        }
        Console.WriteLine(args.Command.CommandText);
    }
});

var p = new SyncParameters(("Name", "Bikes"));
var s = await agent.SynchronizeAsync("filterproducts", p);

```
As you can see, we are executing stored procedures from both scopes, depending on the command type we need.

Since it required some code modification, you can test it on the branch https://github.com/Mimetis/Dotmim.Sync/tree/OnGetCommand

Let me know if it can solve your problem

**Reply by slagtejn** [reply]
This might solve the issue but still creating unnecessary stored procs per scope.  If there are scopes filterproducts_1, filterproducts_2...etc, the _changes & _initialize stored procs would be duplicated and the only difference is the name.

I was thinking of a solution where you would pass in scopeInstanceName to the SynchroizeAsync function along with scopeName.  In your example the “default” scope would also contain the filters.  So in your example it would be SynchroizeAsync("default", "filterProducts", p).  By default the scopeInstanceName would be the same as the scopeName so backwards compatibility wouldn't be an issue.  The scopeInstanceName is really just needed to track the meta data (server/client timestamps, duration, sync times) which is stored in the scope_info_history table on the server and the scope_info table on the client.  The sync_scope_schema & sync_scope_setup is still based on the scopeName.    The scope_info_server table would not contain the scopeInstanceName's, but the scopeNames like it does now.  The client scope_info table would contain scopeInstancNames, but the  sync_scope_schema/sync_scope_setup would be from the scopeName.  What do you think?  This solution you would be able to track data(timestamps) on a scope based on the instance name.

**Reply by Mimetis** [reply]
You don't need to generate `_changes` & `_initialize` stored procedures if filter is the same (and only value is changing)
You can complete the sample and redirect any "value" to the good stored procedure.

for your scopeInstanceName  and scopeName, I guess a schema would be helpful to understand what you are trying to achieve here

**Comment by cdash** [comment]
I think I have a good idea of how it could be accomplished but it definitely would require some schema changes. The changes are fairly minor on the server-side and a little larger on the client-side but it would end up making the two schemas more uniform in structure.

The key thing that this would seem to allow is being able to dynamically add new filters on an as needed basis without having to reinitialize, causing all the data to be downloaded again.

Think of a person using an application that needs to synchronize all the asset data for the city they are currently working in. They have a few main cities they work out of but they also sometimes need to travel to new locations. This would allow them to have a scope instance per city and when they go to a new location a new instance is created with the new city parameter. Then, only the assets for that city will be downloaded, keeping all the asset data for their main cities still in tact. In fact the data for those other cities can still continue syncing and receiving updates so that when they get back everything is already in sync.

```sql
CREATE TABLE [dbo].[scope_info_history](
	[sync_scope_id] [uniqueidentifier] NOT NULL, -- Actually the client id
	[sync_scope_name] [nvarchar](100) NOT NULL, -- Name of the scope config
        [sync_scope_instance] [nvarchar](100) NOT NULL -- Name of the instance (default to sync_scope_name)
	[scope_last_sync_timestamp] [bigint] NULL,
	[scope_last_sync_duration] [bigint] NULL,
	[scope_last_sync] [datetime] NULL,
)

-- Client only has a single scope_info table so probably it would make more sense if it were split into two tables like the server is

CREATE TABLE [dbo].[scope_info](
	[sync_scope_name] [nvarchar](100) NOT NULL,
	[sync_scope_schema] [nvarchar](max) NULL,
	[sync_scope_setup] [nvarchar](max) NULL,
	[sync_scope_version] [nvarchar](10) NULL,
)

CREATE TABLE [dbo].[scope_info_history](
    [sync_scope_id] [uniqueidentifier] NOT NULL, -- Actually the client id
    [sync_scope_name] [nvarchar](100) NOT NULL, -- Name of the scope config
    [sync_scope_instance] [nvarchar](100) NOT NULL -- Name of the instance (default to sync_scope_name)
    [scope_last_server_sync_timestamp] [bigint] NULL,
    [scope_last_sync_timestamp] [bigint] NULL,
    [scope_last_sync_duration] [bigint] NULL,
    [scope_last_sync] [datetime] NULL,
)
```

```c#
// Creating one Setup
// This Setup contains only the assets table
var setup = new SyncSetup("Assets");

// Create a filter On the city column of the assets tables
var assetsFilter = new SetupFilter("Assets");
productFilter.AddParameter("City", DbType.String);
productFilter.AddWhere("City", "Assets", "City");
setup.Filters.Add(assetsFilter);

var serverProvider = new SqlSyncProvider(DBHelper.GetDatabaseConnectionString(serverDbName));
var clientProvider = new SqlSyncProvider(DBHelper.GetDatabaseConnectionString(clientDbName));

// Creating an agent that will handle all the process
var agent = new SyncAgent(clientProvider, serverProvider);

var parameters1 = new SyncParameters
{
    { "City", "New York" },
};

var parameters2 = new SyncParameters
{
    { "City", "Los Angeles" },
};

// Sync the scope providing an instance name of Assets_NewYork. The name doesn't really matter as long as it is unique and is always supplied the same set of parameters.
var s1 = await agent.SynchronizeAsync(setup, parameters1, "Assets_NewYork);


// Sync the scope providing an instance name of Assets_LosAngeles. The name doesn't really matter as long as it is unique and is always supplied the same set of parameters.
var s1 = await agent.SynchronizeAsync(setup, parameters2, "Assets_LosAngeles);

// This will result in two records being stored in the scope_info_history table.
```

**Reply by Mimetis** [reply]
If I understand well,  the scope_name_instance is basically the different parameters values ?

**Reply by cdash** [reply]
Yes, in this case it maps 1:1 to each unique set of parameters as the goal is to have a tracking of timestamps for each set of parameters. It's possible that it could be used in some other scenario but honestly, I can't think of one at the moment. I think your solution above was trying to define multiple types of filtering for a single set of tables but in this scenario, it is always the same filter and same scope config but just different sets of parameters. The creation of multiple scopes was being done originally because it is not currently possible without doing so.

It would also be possible that instead of specifying the instance name manually that it is automatically calculated using some type of hash of the parameter set. If that were done then we would still need some type of toggle on whether it should create an instance for each set of parameters or always map to the default in case this functionality does not want to be used.

**Reply by Mimetis** [reply]
> It would also be possible that instead of specifying the instance name manually that it is automatically calculated using some type of hash of the parameter set

I was indeed thinking of something like that.
Getting and saving a hash of the `SyncParameters` instance. 
We can easily order the parameters object and gets the same hash every time, until you change the values of the parameters.

I still see some weird side effects

Here is the first one I found so far:

- You are synchronizing `{ "City", "Los Angeles" }`
- You are synchronizing then `{ "City", "New York" },`
- Your database client is up to date with all the cities rows for both city.
- Later, you are out of scope for any reason and need to resync everything (whatever the reason tbh)
- You are doing a `Reinitialize` (or `ReinitializeWithUpload`) with the parameters `{ "City", "Los Angeles" }`
- You are up to date with `{ "City", "Los Angeles" }`,
- You lost everything related to `{ "City", "New York" }`
- Fine you decide to do the same `Reinitialize` with `{ "City", "New York" }`
- You are good to go with `{ "City", "New York" }` but this time you lost everything related to `{ "City", "Los Angeles" }`

For now the only solution I found is to implement a new behaviour for **Reiniatilize**:

- `ReinitializeWithReset` : Reset all the tables then make a reinitialize
- `ReinitializeWithoutReset`: Reinitialize everything but without resetting the tables
- `ReinitializeWithUploadWithReset` and `ReinitializeWithUploadWithoutReset`

I start to become more and more complex in my opinion...

**Reply by cdash** [reply]
Instead of creating new options for the **Reinitialize** wouldn't we be able to just use the filter parameters themselves to control which records get deleted from the table?

**Reply by Mimetis** [reply]
Only if your `SyncSetup` contains only tables with filters
But if you have a table that is part of your `SyncSetup`, but where there is no filters, it won't work

`Reinitialize` is here to start from scratch again, basically.

**Comment by slagtejn** [comment]
I will test the new OnGetCommand interceptor since this seems like the solution with the least amount of changes and won't cause backwards compatibility issues.

In the scenario of the creating new scopes on the client, it will automatically try to provision on the client/server.  We can block the entire provision process through this interceptor?

**Comment by Mimetis** [comment]
Hey @slagtejn, @cdash,

I've been working for the last 5 days on a new version of `DMS` to match your scenario.
It's a huge upgrade for the framework, and I had to change a lot of things.
Honestly, I was not convinced it was worth it, but now I'm pretty excited about it !

From your perspective, the most important to notice is that now, we have 2 tables on each side (server / client) that are identical:

- `scope_info`: This table now contains only the **setup** and **schema**. Each line is identified by the `scopeName` property
- `scope_info_client`: This table contains all clients that needs to synchronize a setup from `scope_info`. 
Each client is identified by the `scopeName` property and the `syncParameters` used

Those two tables are created on both side and are identical. 
No more `scope_info_server` on server side (both sides have now a `scope_info` table)
No more `scope_info_history` on server side (both sides have now a `scope_info_client` table)

I realized that doing this made the DMS framework a little bit more _fluent_.. and easier to maintain.

As an example, I've created a sync with 3 tables, 2 of them are filtered (`Product` and `ProductCategory`), the 3rd table does not have any filter (`Employee`)

Here is a `scope_info` containing a setup with 3 tables:
- `ProductCategory`
- `Product`
- `Employee`

<img width="752" alt="image" src="https://user-images.githubusercontent.com/4592555/187087016-fb28422c-a829-4f42-a35b-69de553977d2.png">

As you can see, we don't have any more the last sync timestamp information here.
The new `scope_info_client` contains 2 lines:
- First one is for the filter `ProductCategoryId == 'MOUNTB'`
- Second one is for the filter `ProductCategoryId == 'ROADFR`

Each `scope_info_client` record is totally independent and contains its own Last Sync Timestamp (server and client)

<img width="926" alt="image" src="https://user-images.githubusercontent.com/4592555/187087045-deda1eef-ac95-430d-87e9-0e5cb7ed3175.png">


Here is the code I used to have this result:

``` cs
var setup = new SyncSetup("ProductCategory", "Product", "Employee");
setup.Tables[productCategoryTableName].Columns.AddRange("ProductCategoryId", "Name", "rowguid", "ModifiedDate");
setup.Filters.Add("ProductCategory", "ProductCategoryId");
setup.Filters.Add("Product", "ProductCategoryId");


var pMount = new SyncParameters(("ProductCategoryId", "MOUNTB"));
var pRoad = new SyncParameters(("ProductCategoryId", "ROADFR"));

var agent = new SyncAgent(client.Provider, Server.Provider);
var r1 = await agent.SynchronizeAsync("v1", setup, pMount);
var r2 = await agent.SynchronizeAsync("v1", setup, pRoad);
```

It's a big step moving forward for `DMS` as I never imagined to be able to sync a client database with different parameters value
Let me know what you think.

If you want to test this version, you can grab the branch [ClientHistory](https://github.com/Mimetis/Dotmim.Sync/tree/ClientHistory)

Be careful, there is no upgrade path from the current release version to this new one.
Please test it on a new database, and not in production :)

And please let me know what you think about it

**Reply by slagtejn** [reply]
This is amazing and really makes DMS powerful when using filters! I like that the schemas are unified now.  I will test this solution with a new setup.

**Reply by cdash** [reply]
Just took a look at, it is looking very nice. You even implemented the automatic hashing of the parameters and included the actual values in the database on the server side so you can easily see which values map to each instance.

Looks like it was quite a bit of work, but it definitely seems like it will simplify things moving forward by having the client and server schemas exactly the same.

**Comment by slagtejn** [comment]
I tested the branch and it's working good!  Table structure is clean.  Huge improvement to DMS.

**Reply by Mimetis** [reply]
Did you test to migrate from an old version to the new one ? #799
